federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,244 @@ limitations under the License.
+			return zone, nil
+		}
+	}
+	return nil, fmt.Errorf("DNS zone %s not found.", dnsZoneName)
+}
+
+/* getRrset is a hack around the fact that dnsprovider.ResourceRecordSets interface does not yet include a Get() method, only a List() method.  TODO:  Fix that.
+   Note that if the named resource record set does not exist, but no error occurred, the returned set, and error, are both nil
+*/
+func getRrset(dnsName string, rrsetsInterface dnsprovider.ResourceRecordSets) (dnsprovider.ResourceRecordSet, error) {
+	var returnVal dnsprovider.ResourceRecordSet
+	rrsets, err := rrsetsInterface.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, rrset := range rrsets {
+		if rrset.Name() == dnsName {
+			returnVal = rrset
+			break
+		}
+	}
+	return returnVal, nil
+}
 
+/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resource record sets for dnsName are consistent with endpoints.
+   if endpoints is nil or empty, a CNAME record to uplevelCname is ensured.
+*/
+func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoints []string, uplevelCname string) error {
+	dnsZone, err := getZone(dnsZoneName, s.dnsZones)
+	if err != nil {
+		return err
+	}
+	rrsets, supported := dnsZone.ResourceRecordSets()
+	if !supported {
+		return fmt.Errorf("Failed to ensure DNS records for %s. DNS provider does not support the ResourceRecordSets interface.", dnsName)
+	}
+	rrset, err := getRrset(dnsName, rrsets) // TODO: rrsets.Get(dnsName)
+	if err != nil {
+		return err
+	}
+	if rrset == nil {
+		// It doesn't exist yet, so create it
+		if len(endpoints) < 1 {
+			// There are no endpoint addresses at this level, so CNAME to uplevel, if provided
+			if uplevelCname != "" {
+				newRrset := rrsets.New(dnsName, []string{uplevelCname}, minDnsTtl, rrstype.CNAME)
+				rrset, err = rrsets.Add(newRrset)
+				if err != nil {
+					return err
+				}
+			}
+			// We want no record, and we have no record, so we're all good.
+		} else {
+			// We have valid endpoint addresses, so just add them as A records.
+			// But first resolve DNS names, as some cloud providers (like AWS) expose
+			// load balancers behind DNS names, not IP addresses.
+			resolvedEndpoints := make([]string, len(endpoints))
+			for _, endpoint := range endpoints {
+				if net.ParseIP(endpoint) == nil {
+					// It's not a valid IP address, so assume it's a DNS name, and try to resolve it,
+					// replacing it's DNS name with it's IP addresses in expandedEndpoints
+					if ipAddrs, err := net.LookupHost(endpoint); err != nil {
+						return err
+					} else {
+						resolvedEndpoints = append(resolvedEndpoints, ipAddrs...)
+					}
+				} else {
+					resolvedEndpoints = append(resolvedEndpoints, endpoint)
+				}
+			}
+			newRrset := rrsets.New(dnsName, endpoints, minDnsTtl, rrstype.A)
+			rrset, err = rrsets.Add(newRrset)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		// the rrset already exists, so make it right.
```

Comment:
No, these change.  When they do, the watches against the status of the underlying kubernetes services trigger, and this code gets called, which updates the DNS records accordingly.  Make sense?